lib/organization.js

```diff
@@ -273,4 +271,97 @@ export class Organization {
     // NB: this will not work for proposals using votingMachine's that are not the default one
     return this.votingMachine.vote(proposalId, choice, params);
   }
+
+  /**
+   * Returns global constraints currently registered into this DAO, as Array<OrganizationGlobalConstraintInfo>
+   * @param name like "TokenCapGC"
+   */
+  async globalConstraints(name) {
+    // return the global constraints registered on this controller satisfying the contract spec
+    // return all global constraints if name is not given
+    const constraints = await this._getConstraints();
+    if (name) {
+      return constraints.filter(s => s.name === name);
+    } else {
+      return constraints;
+    }
+  }
+
+  /**
+   * returns global constraints currently in this Organization, as OrganizationGlobalConstraintInfo
+   */
+  async _getConstraints() {
+    // private method returns all registered schemes.
+    // TODO: this is *expensive*, we need to cache the results (and perhaps poll for latest changes if necessary)
+    const constraintsMap = new Map(); // <string, OrganizationGlobalConstraintInfo>
+    const controller = this.controller;
+    const avatar = this.avatar;
+    const arcTypesMap = new Map(); // <address: string, name: string>
+    const contracts = await getDeployedContracts();
+
+    /**
+     * TODO:  This should pull in all known versions of the schemes, names
+     * and versions in one fell swoop.
+     */
+    for (const name in contracts.allContracts) {
+      const contract = contracts.allContracts[name];
+      arcTypesMap.set(contract.address, name);
+    }
+
+    const event = controller.AddGlobalConstraint(
+      {},
+      { fromBlock: 0, toBlock: "latest" }
+    );
+
+    await new Promise(resolve => {
+      event.get((err, eventsArray) =>
+        this._handleConstraintEvent(
+          err,
+          eventsArray,
+          true,
+          arcTypesMap,
+          constraintsMap
+        ).then(() => {
+          event.stopWatching();
+          resolve();
+        })
+      );
+    });
+
+    const registeredConstraints = [];
+
+    for (const gc of constraintsMap.values()) {
+      if (await controller.isGlobalConstraintRegister(gc.address, avatar.address)) {
+        registeredConstraints.push(gc);
+      }
+    }
+
+    return registeredConstraints;
+  }
+
+  async _handleConstraintEvent(
+    err,
+    eventsArray,
+    adding,
+    arcTypesMap,
+    constraintsMap // : Promise<void>
+  ) {
+    if (!(eventsArray instanceof Array)) {
+      eventsArray = [eventsArray];
+    }
+    const count = eventsArray.length;
+    for (let i = 0; i < count; i++) {
+      const address = eventsArray[i].args._globalconstraint;
+      const paramsHash = eventsArray[i].args._params;
+
+      const info = {
+        address: address,
+        paramsHash: paramsHash,
+        // will be undefined if not a known scheme
+        name: arcTypesMap.get(address)
+      };
+
+      constraintsMap.set(address, info);
```

Comment:
why not just check if its registered here instead of adding all of them to the Map and then iterating through the whole thing again to see which ones are registered?

Reply:
the map eliminates dups

Reply:
commented

Reply:
so you think its more likely that there will be dupes to eliminate first rather than unregistered constraints to eliminate first? either way its not a big deal